Add resource management alerts (#82)
appuio/component-openshift4-monitoring: docs/modules/ROOT/pages/explanations/resource_management.adoc

)
<
// The pod capacity of the largest worker node
max(
  kube_node_status_capacity{resource="pods"}
    * on(node) group_left kube_node_role{role="app"}
)
```

.**Expect too many pods**

We alert if we predict to be unable to run all currently running pods in three days if the largest worker node crashes.
```
// The pod capacity of all worker nodes
sum(
  kube_node_status_capacity{resource="pods"}
    * on(node) group_left kube_node_role{role="app"}
)
-
// The predicted number of running pods on all worker nodes in 3 days
predict_linear(
  // Take the moving average over a day to prevent flapping alerts
  avg_over_time(
    sum(
      kubelet_running_pods
        * on(node) group_left kube_node_role{role="app"},
    )[1d:1h]
  )[1d:1h],
3*24*60*60)
<
// The pod capacity of the largest worker node
max(
  kube_node_status_capacity{resource="pods"}
    * on(node) group_left kube_node_role{role="app"}
)
```

=== Memory/CPU requests

The combined memory and CPU requests of all workloads is another good indicator.
If the combined requests are higher than the actual resources, some workloads will no be scheduled.

The component provides four resource request related alerts.

> They alert if there isn't enough CPU or memory capacity if the largest worker node disappears or if we predict that the cluster will exceed this threshold in three days.,

Reviewer: ```suggestion
They alert if there isn't enough CPU or memory capacity if the largest worker node disappears or if we predict that the cluster will exceed this threshold in three days.
```

.**Workers CPU usage high**
We alert if there is fewer idle CPU cores than the largest worker node has.
```
sum(
  // The average number of idle CPUs over 15 minutes for all worker nodes
  rate(node_cpu_seconds_total{mode="idle"}[15m])
    * on(instance) group_left label_replace(kube_node_role{role="app"}, "instance", "$1", "node", "(.+)"))
<
// The capacity of the largest worker node
max(
  kube_node_status_capacity{resource="cpu"}
    * on(node) group_left kube_node_role{role="app"}
)
```

.**Workers CPU usage expected to be high**
We alert if we predict to have fewer idle CPU cores than the largest worker node has in three days.
```
// The predicted number idle CPUs for all worker nodes in 3 days
predict_linear(
  // Take the moving average over a day to prevent flapping alerts
  avg_over_time(
    sum(
      rate(node_cpu_seconds_total{mode="idle"}[15m])
        * on(instance) group_left label_replace(kube_node_role{role="app"}, "instance", "$1", "node", "(.+)")
    )[1d:1h]
  )[1d:1h],
3*24*60*60)
<
// The capacity of the largest worker node
max(
  kube_node_status_capacity{resource="cpu"}
    * on(node) group_left kube_node_role{role="app"}
)
```

[NOTE]
====

> By default for predictions we take the one day moving average over one day to predict three days into the future.

Reviewer: ```suggestion
By default, we use the one day moving average over one day for queries which predict three days into the future.
```

Without the moving average the prediction is influenced too much by temporary changes.

> If for example an environment is updated using a blue-green deployment, without the moving average the alert will see this sudden increase in resource usage, interpolate this for three days and will fire.

Reviewer: ```suggestion
For example, if an environment is updated using a blue-green deployment, without the moving average the alert will see this sudden increase in resource usage, extrapolate the increase over three days and will fire, generating a false-positive alert.
```

The moving average gives us a better indication of the long term trend.
====

== Non-Indicators

There are some metrics that might be considered as an indicator for cluster capacity, but have intentionally not been added as alert rules, as they're either too noisy or not actionable.

.**Memory/CPU limits**

Similarly to the total memory and CPU requests of workloads one could look at the total memory and CPU limits as an indicator for cluster capacity.
However in almost all cases the total limits are a lot higher than the actual capacity of the cluster.
This is normal and this overprovisioning is one of the advantages of Kubernetes.

> It0s hard to say what level of overprovisioning is OK and what's not, so observing the actual resource usage is more effective.

Reviewer 2: ```suggestion
It's hard to say what level of overprovisioning is OK and what's not, so observing the actual resource usage is more effective.
```
